### PR TITLE
Self managing allocations

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

**Comment by oli-obk** [comment]
r? @RalfJung 

pulled out of https://github.com/rust-lang/rust/pull/55260 to keep that PR simpler

**Review by RalfJung — CHANGES_REQUESTED** [review]
I expect that the miri submodule update will get messy because we don't know in which order our PRs will land. As long as we have so many miri-breaking PRs in flight, I think it is better to do the submodule updates separately.

File: src/librustc/mir/interpret/allocation.rs [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+        self.undef_mask.set_range(
+            ptr.offset,
+            ptr.offset + size,
+            new_state,
+        );
+        Ok(())
+    }
+}
+
+impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: I thought I kept these things in order :/

File: src/librustc/mir/interpret/allocation.rs [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Hash, RustcEncodable, RustcDecodable)]
+pub struct Allocation<Tag=(),Extra=()> {
+    /// The actual bytes of the allocation.
+    /// Note that the bytes of a pointer represent the offset of the pointer
+    pub bytes: Vec<u8>,
+    /// Maps from byte addresses to extra data for each pointer.
+    /// Only the first byte of a pointer is inserted into the map; i.e.,
+    /// every entry in this map applies to `pointer_size` consecutive bytes starting
+    /// at the given offset.
+    pub relocations: Relocations<Tag>,
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: It's not trivial at least. There is the `copy` method on `Memory`, which requires access to two Allocations. Everything gets rather complicated in there.

File: src/librustc/mir/interpret/mod.rs [hunk 3]
```diff
@@ -523,62 +525,6 @@ impl<'tcx, M: fmt::Debug + Eq + Hash + Clone> AllocMap<'tcx, M> {
...
-            mutability: Mutability::Mutable,
-            extra: Extra::default(),
-        }
-    }
-}
-
-impl<'tcx> ::serialize::UseSpecializedDecodable for &'tcx Allocation {}
-
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, RustcEncodable, RustcDecodable)]
 pub struct Relocations<Tag=(), Id=AllocId>(SortedMap<Size, (Tag, Id)>);
```
Comment (RalfJung) [inline src/librustc/mir/interpret/mod.rs]: `Relocations` seems to also better fit into `alloc.rs`?

Comment (oli-obk) [reply src/librustc/mir/interpret/mod.rs]: yea, I decided to move as little as possible, but doing so in a separate commit should be manageable

File: src/librustc/mir/interpret/allocation.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: I think there's a way to do this, but not right now. I have a few refactorings planned around `AllocId` (that will not do the stupid stuff you saw in the other PR) that will make this possible in 100% safe Rust, without code duplication and without performance regressions

[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: Okay, let's do that later.

File: src/librustc/mir/interpret/mod.rs (hunk 3 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/mod.rs]: And what about `Pointer`? I am consistently confused to find it in `mod.rs`.

**Comment by oli-obk** [comment]
before removing the size==0 checks I was seeing nice perf improvements. Lets see if this was me failing to run perfruns properly.

@bors try

**Comment by oli-obk** [comment]
@rust-timer build 80add95c736112b8f7890dfb4a7cfb854fc9235e

File: src/librustc/mir/interpret/allocation.rs [hunk 4]
```diff
@@ -198,7 +198,6 @@ impl<'tcx, Tag: Copy, Extra: AllocationExtra<Tag>> Allocation<Tag, Extra> {
         align: Align,
         check_defined_and_ptr: bool,
     ) -> EvalResult<'tcx, &[u8]> {
-        assert_ne!(size.bytes(), 0, "0-sized accesses should never even get a `Pointer`");
```
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: I'm fine with keeping the assertions. I am also fine with removing them.

The important part is that we don't short-circuit for size zero.

**Comment by RalfJung** [comment]
Do we also want to find a new home for `Pointer`? `mod.rs` is inadequate, I think. I just cannot decide between `value.rs` and `alloc.rs`.

File: src/librustc/mir/interpret/allocation.rs (hunk 4 above)
Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: I can turn them into debug assertions. There's a CI runner with debug assertions on.

Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: But I checked, the code works fine even if given zero sized args

**Comment by oli-obk** [comment]
> Do we also want to find a new home for Pointer? mod.rs is inadequate, I think. I just cannot decide between value.rs and alloc.rs.

pointer.rs ?

File: src/librustc/mir/interpret/allocation.rs (hunk 4 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: I am fine either way.

**Comment by RalfJung** [comment]
What would be in there? `Pointer` and the pointer arithmetic stuff?

Sounds reasonable.

File: src/librustc/mir/interpret/allocation.rs (hunk 4 above)
Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: I'll leave them out. No need to check an invariant that doesn't actually exist

**Comment by RalfJung** [comment]
r=me if you remove the miri update and perf looks good

**Comment by RalfJung** [comment]
> Not fetching those allocations 4x for 4 checks does have an effect.

Where did that happen?

**Comment by RalfJung** [comment]
Oh I see... the `get_bytes` methods used to hit the alloc hash map *a lot*.

Nice :)

**Comment by pietroalbini** [comment]
@bors retry

Prioritizing beta release.

**Comment by kennytm** [comment]
@bors retry

Failed to remove a file due to "Access Denied". Very likely spurious.

**Comment by RalfJung** [comment]
Again a permission problem, accessing `dmesg`? Cc @rust-lang/infra 

@bors retry

**Comment by oli-obk** [comment]
Rustdoc failed to compile with `rustc exited with signal: 10`?!

**Comment by kennytm** [comment]
Same error again, likely legit.

**Comment by oli-obk** [comment]
I am unable to reproduce on x86 linux. I also don't see how these changes could cause changes in the emitted machine code. It's just a refactoring moving functions between modules.

I could do this PR in steps by splitting it into smaller PRs

**Comment by RalfJung** [comment]
I don't have any better idea either. Maybe reduce this one to the most bitrotty "moving stuff around" part (`Allocation`, `ScalarMaybeUndef`, `Pointer`), and then have a separate PR for actually adding more methods to `Allocation`?

**Comment by oli-obk** [comment]
blocked on https://github.com/rust-lang/rust/pull/55674

**Comment by oli-obk** [comment]
closing in favour of #55915

**Pull request closed without merging**